Add script to automate setup of input files and folders

davereinhart opened the pull request:
Adds a new script to automate preparation of files needed for the `create_submissions` script. This includes QC and standardization of metadata Excel file, extracting FASTA and metrics files, querying metadata from LIMS or ID3C, processing with NextClade and VADR, identifying records to exclude based on data quality, pulling list of previous submissions, preparing and exporting the necessary files.

davereinhart committed:
Add script to automate setup of input files and folders

The initial steps in this process involves extracting, creating, and renaming
files and folders. This automates those steps leading up to the first manual step
in the process (uploading the FASTA .fa file to NexClade).

davereinhart committed:
Add NextClade processing

Download sars-cov-2 data from NextClade and process FASTA file to generate
tsv output.

davereinhart committed:
Determine excluded VOCs from NextClade output

Apply rules to determine excluded VOCs and save NWGC IDs to excluded-vocs.txt file.

davereinhart committed:
Add VADR processing to file preparation script

To help futher consolidate sequencing submission workflow, adding processing of FASTA file with VADR
to this script. This requires that Docker is installed and currently running. Also included an option to skip
this step, which can be run later via `/submissions/script/run_vadr` script if needed.

davereinhart commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Automate pieces of sequencing submission pipeline](https://trello.com/c/eLaRZmyh/1420-automate-pieces-of-sequencing-submission-pipeline)

davereinhart committed:
Add QC of external metadata Excel file

Standardize and validate barcode and collection date values in the external metadata
Excel file. The Excel file contains 2 sheets which are loaded to dataframes, then written back
to Excel after minor transformations and passing all validation checks.

Barcodes are validated as 8-digit hexidecimal values. Collection dates are updated to remove
"na" values and confirm no dates prior to 2020-02-01 and that not all values are identical.

davereinhart committed:
Create CSV of SFS sample barcodes

Automate creation of `sfs-sample-barcodes.csv` as part of this process, rather
than running it as a separate script.

davereinhart committed:
Identify and optionally drop expirimental samples

Expirimental samples have barcodes ending in `_exp` and typically can be removed
after confirming with sequencing team. This change displays those records and prompts
for user input to optionally drop them before saving back to Excel.

davereinhart committed:
Identify and optionally drop Cascadia samples

Cascadia samples will need to be handled differently, since ID3C does not have
any information about these samples. Until this process is updated to handle these records,
adding an option to drop them from the metadata file during the file prep process.

davereinhart committed:
Refactoring

Combining user input prompts into a generic function to reduce repetitive code.

davereinhart committed:
Remove extra directory created by file prep script

Initial code created two separate directores (`Batch-YYYYMMDD` and `YYYYMMDD_fastq`), but on further review of archived
submission files, only one directory is needed.

davereinhart committed:
Run `export_id3c_metadata` script as part of file prep

Optionally run `export_id3c_metadata` bash script while running the file
prep script to avoid running as a separate command.

This temporarily redirects stdout to a file `id3c_metadata_with_county.csv` and
saves to the output batch directory. The file is loaded to a dataframe and checked
to make sure all records with source `SFS` have a collection date, with an option
to either continue or bail out if any are missing.

davereinhart committed:
Download previous submission identifiers from Github

Adding option to download `hcov19-sequencing-identifiers.tsv` from Github and
save to output batch directory as `previous-submissions.tsv`, eliminating another
manual step from the workflow.

davereinhart committed:
Metadata QC check for matching sample IDs and barcodes

The Excel file containing external metadata contains two sheets which should have
matching lists of sample IDs and barcodes. Adding a QC check to confirm before proceeding.
If there is a mismatch between the two sheets, it should be corrected in the source file
before re-running the `prep_files` script.

davereinhart committed:
Validate expected outputs that are required to create submission files

Consolidate output paths to a single dictionary and check that all exist at the
end of the script. If none are missing, prints a pre-formatted command to run
`create_submissions.py` with the appropriate arguments, otherwise a list of the
missing files or folders.

davereinhart committed:
Calculate next available strain ID from previous submissions file

The next available strain ID is a required argument for the `create_submissions.py`
script. This change will calculate that value based on contents of the previous
submissions and include it in the pre-formatted command that is output at the end.

davereinhart committed:
Code changes to process older batches

Until now, the code had only been tested with the most recent two sequencing batches.
Older batches contained slightly different formats which are being handled with these
changes. Also adding code to only standardize internally generated barcodes, since identifiers
for external samples have multiple formats.

davereinhart committed:
Get LIMS metadata

Adding module to pull metadata from LIMS API to be used in file prep script. LIMS will be used
as the metadata source going forward, in place of ID3C.

davereinhart committed:
Updates for OR DOH reporting, LIMS API for metadata, author lists

Add CSV output for Oregon DOH reporting based on template they provided. Pull additional
patient metadata from LIMS API to include in this output.

Updates to author lists.

davereinhart commented:
WIth `sfs-submissions` conda env activated, this command will run the prep_files script. The tar.gz file can be downloaded from Globus and the associated external metadata Excel file is posted in the `data-transfer-nwgc` channel.
```
LOG_LEVEL=debug python ./submissions/scripts/prep_files.py \
--results-file <path to tar.gz file> \
--metadata-file <path to external metadata XLSX> \
--output-dir <output dir> \
--batch-date <YYYYMMDD>
```

These environment variables are required for pulling metadata from LIMS API: `LIMS_API_KEY_ID`, `LIMS_API_KEY_SECRET`, `LIMS_API_URL` (if pulling from ID3C, set `PGSERVICE=seattleflu-production`)

If the script succeeds, it will print a formatted command for running the `create_submissions` script.

bencap commented on submissions/scripts/prep_files.py:
```diff
@@ -0,0 +1,460 @@
+    else:
+        raise ValueError(f"Identifiers must be hexidecimal barcodes, `blank`, or `twist positive`. Invalid barcode: {standardized_barcode}")
+
+def validate_collection_dates(collection_dates):
```
Not sure if there have been any, but it may also be worth validating dates are not in the future given some of the errors we have seen with collection dates in the past.

davereinhart replied:
Good idea, I'll add that in!

bencap commented on submissions/scripts/prep_files.py:
```diff
@@ -0,0 +1,460 @@
+        return None
+    elif isinstance(barcode, str):
+        standardized_barcode = barcode.strip().lower()
+    elif isinstance(barcode, int):
```
If this works I assume you haven't encountered any barcode floats? Sometimes Pandas casts to floats depending on the barcode string so that may pop up in the future.

davereinhart replied:
Haven't seen that happen yet, but I'll keep an eye out for it. When that does occur, can it be cast to str or does it need reformatting?

bencap replied:
I want to say you have to be extra careful with them because its primarily barcodes with the format `123456e9` that convert to exponents. With that said, I haven't seen it happen with new new Excel reader. Mostly just a warning I guess, this code will catch it if it comes up.

bencap commented on envs/submissions.yaml:
```diff
@@ -7,4 +7,7 @@ dependencies:
   - pandas
+  - conda
+  - nextclade
+  - docker-py
```
Did a fresh install and also needed to install `xlsxwriter`.

bencap commented on submissions/scripts/export_lims_metadata.py:
```diff
@@ -0,0 +1,190 @@
+    print(f"Received {len(result)} additional metadata (Oregon) records from LIMS.")
+
+    output_df = pd.merge(input_df, result, on=['sfs_collection_barcode'])
+    #remaining_records = input_df[(~input_df.nwgc_id.isin(output_df.nwgc_id))]
```
Are these commented lines used on occasion?

davereinhart replied:
For now, the only OR samples are from Cascadia, and it should always be matching on collection barcode for that project, so the follow-up join on sample barcodes isn't necessary.

bencap commented on submissions/scripts/prep_files.py:
```diff
@@ -0,0 +1,460 @@
+    if not exp_samples.empty:
+        print(f"Expirimental samples found: {len(exp_samples)}\n {exp_samples.to_string()}")
+
+        if yes_no_cancel("Drop expirimental samples?"):
```
Out of curiosity, are these experimental samples usually dropped?

davereinhart replied:
Yes, these are dropped after confirming with the sequencing team that they are cost center related. Otherwise, they may need to be included.

bencap commented on submissions/scripts/export_lims_metadata.py:
```diff
@@ -0,0 +1,190 @@
+            'specimen.ids.collectionBarcode': 'sfs_collection_barcode',
+            'specimenCollectionSite': 'specimen_collection_site'
+        }
+        df.rename(columns=column_map, inplace=True)
```
Might want to add a check here that the df is non-empty before trying to rename, since if you try to rename an empty dataframe the column names won't show up and you'll get an index error down the line. This was happening to me with the 8-10 batch. Something like the code below is what I was using to get it working:
```
        if df.empty:
            df = pd.DataFrame(columns=column_map.values())
        else:
            df.rename(columns=column_map, inplace=True)
```

davereinhart replied:
Thanks for catching that! Looks like that should only occur when there are no Oregon samples, in which case we can skip the call to `add_lims_additional_metadata` entirely.  I'll make that change.

bencap reviewed (commented):
This seems like it will be a lot easier now with this script, thanks for putting it together!

davereinhart committed:
Skip OH DOH report if no valid samples, prevent future dates

Changes based on PR review. Skipping OR DOH report if no valid samples are found
and preventing future collection dates.

Added missing library to conda environment.

bencap reviewed (approved):
Thanks for the changes, looks gtg now!

davereinhart committed:
Populate sfs_sample_identifier for Cascadia samples

Populate `sfs_sample_identifier` with the `sfs_identifier_for_doh_reporting` value retrieved
from LIMS for Cascadia project.

davereinhart commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Sequencing submission to OR HA](https://trello.com/c/QpvuIxgQ/1433-sequencing-submission-to-or-ha)
